File: Trabajo pasado/src/icc/ajedrez/Caballo.java
package icc.ajedrez;
import java.util.List;
import java.util.ArrayList;

/**Clase para el caballo del ajedrez.
*@author Sebastián A.
*/
public class Caballo extends Pieza {

	/**Método constructor para el caballo.
	*@param columna Columna en la que irá la pieza.
	*@param renglón Renglón en el que irá la pieza.
	*@param color Color de la pieza.
	*/
	public Caballo(char columna, int renglón, boolean color) {
		super(columna, renglón, color);
	} //Fin de constructor Caballo.

	/**Método para saber todos los posibles movimientos de la pieza.
	*@return Lista con los posibles movimientos de la pieza en la forma (columna,renglón).
	*/
	public List posiblesMovimientos() {

		int posibles = 0;
		int renglónContador;
		char columnaContador;
		char getColumna;
		List<String> listaDePosibles = new ArrayList<String>();

		//NORTE->ESTE.
		getColumna = getColumna();
		renglónContador = getRenglón() + 2;
		columnaContador = ++getColumna;
		if (renglónContador < 9 && columnaContador <= 'h') {
			posibles += 1;
			listaDePosibles.add("("+columnaContador+","+renglónContador+")");
		}

		//NORTE->OESTE.
		getColumna = getColumna();
		renglónContador = getRenglón() + 2;
		columnaContador = --getColumna;
		if (renglónContador < 9 && columnaContador <= 'h') {
			posibles += 1;
			listaDePosibles.add("("+columnaContador+","+renglónContador+")");
		}

		//SUR->ESTE.
		getColumna = getColumna();
		renglónContador = getRenglón() - 2;
		columnaContador = ++getColumna;
		if (renglónContador > 0 && columnaContador <= 'h') {
			posibles += 1;
			listaDePosibles.add("("+columnaContador+","+renglónContador+")");
		}

		//SUR->OESTE.
		getColumna = getColumna();
		renglónContador = getRenglón() - 2;
		columnaContador = --getColumna;
		if (renglónContador > 0 && columnaContador <= 'h') {
			posibles += 1;
			listaDePosibles.add("("+columnaContador+","+renglónContador+")");
		}

		//ESTE->NORTE.
		getColumna = getColumna();
		renglónContador = getRenglón() + 1;
		columnaContador = ++getColumna;
		columnaContador = ++columnaContador;
		if (renglónContador > 0 && columnaContador <= 'h') {
			posibles += 1;
			listaDePosibles.add("("+columnaContador+","+renglónContador+")");
		}

		//ESTE->SUR.
		getColumna = getColumna();
		renglónContador = getRenglón() - 1;
		columnaContador = ++getColumna;
		columnaContador = ++columnaContador;
		if (renglónContador > 0 && columnaContador <= 'h') {
			posibles += 1;
			listaDePosibles.add("("+columnaContador+","+renglónContador+")");
		}

		//OESTE->NORTE.
		getColumna = getColumna();
		renglónContador = getRenglón() + 1;
		columnaContador = --getColumna;
		columnaContador = --columnaContador;
		if (renglónContador > 0 && columnaContador <= 'h') {
			posibles += 1;
			listaDePosibles.add("("+columnaContador+","+renglónContador+")");
		}

		//OESTE->SUR.
		getColumna = getColumna();
		renglónContador = getRenglón() - 1;
		columnaContador = --getColumna;
		columnaContador = --columnaContador;
		if (renglónContador > 0 && columnaContador <= 'h') {
			posibles += 1;
			listaDePosibles.add("("+columnaContador+","+renglónContador+")");
		}

		return listaDePosibles;

	} //Fin de posiblesMovimientos().

	/**Método para obtener la pieza representada en forma de String.
	*@return Representación en cadena de la pieza.
	*/
	public String toString() {
		String colorCaballo;
		if (getColor()) {colorCaballo = "blanco";} else {colorCaballo = "negro";}
		return "Este caballo "+colorCaballo+" se encuentra en la posición: ("+getColumna()+","+getRenglón()+").";
	}

} //Fin de clase Caballo.

File: code/piezas/Caballo.rb
require_relative '../Pieza'

# Clase que representa a la pieza Caballo.
class Caballo < Pieza

  # Inicializador de un Caballo.
  # columna:: Columna del Caballo.
  # renglón:: Renglón del Caballo.
  # color:: Color del Caballo (booleano).
  def initialize(columna, renglón, color)
    super(columna, renglón, color);
  end

  def posiblesMovimientos
    # Lista con los posibles movimientos
    # de la pieza.
    listaPM = []

    # En todos los siguientes casos nos
    # moveremos dos espacios hacia la
    # primera dirección y un espacio hacia
    # la segunda dirección.
    # Por ejemplo, NORTE -> ESTE es mover
    # el Caballo dos espacios hacia arriba
    # y uno hacia la derecha.

    # NORTE -> ESTE.
    begin
      x = desplazar(1, 2)
      listaPM.push(x)
    rescue
    end

    # NORTE -> OESTE.
    begin
      x = desplazar(-1, 2)
      listaPM.push(x)
    rescue
    end

    # ESTE -> NORTE.
    begin
      x = desplazar(2, 1)
      listaPM.push(x)
    rescue
    end

    # ESTE -> SUR.
    begin
      x = desplazar(2, -1)
      listaPM.push(x)
    rescue
    end

    # SUR -> ESTE.
    begin
      x = desplazar(1, -2)
      listaPM.push(x)
    rescue
    end

    # SUR -> OESTE.
    begin
      x = desplazar(-1, -2)
      listaPM.push(x)
    rescue
    end

    # OESTE -> NORTE.
    begin
      x = desplazar(-2, 1)
      listaPM.push(x)
    rescue
    end

    # OESTE -> SUR.
    begin
      x = desplazar(-2, 1)
      listaPM.push(x)
    rescue
    end

    # Se regresa la lista...
    return listaPM
  end

  def posiblesMovimientosTablero(tablero)
    # Lista con los posibles movimientos
    # de la pieza.
    listaPM = posiblesMovimientos

    # Se revisa cada posición posible del Caballo en
    # general y se quitan aquellas en donde hay piezas
    # del mismo color que la del caballo.
    for x in listaPM
      pieza = tablero.getPieza(x.columna, x.renglón)
      if pieza && self.color == pieza.color
        listaPM.delete(x)
      end
    end

    # Se regresa la lista...
    return listaPM
  end

end